### PR TITLE
[ThinLTO] Shrink GlobalValueSummary by 8 bytes

### PR DESCRIPTION
**Pull request by kazutakahirata**
During the ThinLTO indexing step for one of our large applications, we
create 7.5 million instances of GlobalValueSummary.

Changing:

  std::vector<ValueInfo> RefEdgeList;

to:

  SmallVector<ValueInfo, 0> RefEdgeList;

in GlobalValueSummary reduces the size of each instance by 8 bytes.
The rest of the patch makes the same change to other places so that
the types stay compatible across function boundaries.

### CODE REVIEW

File: llvm/lib/Analysis/ModuleSummaryAnalysis.cpp [hunk 1]
```diff
@@ -99,10 +99,11 @@ extern cl::opt<bool> MemProfReportHintedSizes;
 // can only take an address of basic block located in the same function.
 // Set `RefLocalLinkageIFunc` to true if the analyzed value references a
 // local-linkage ifunc.
-static bool findRefEdges(ModuleSummaryIndex &Index, const User *CurUser,
-                         SetVector<ValueInfo, std::vector<ValueInfo>> &RefEdges,
-                         SmallPtrSet<const User *, 8> &Visited,
-                         bool &RefLocalLinkageIFunc) {
+static bool
+findRefEdges(ModuleSummaryIndex &Index, const User *CurUser,
+             SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &RefEdges,
```
Comment (teresajohnson) [inline llvm/lib/Analysis/ModuleSummaryAnalysis.cpp:104]: According to https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h (see the "Note" box):

```
Prefer to use ArrayRef<T> or SmallVectorImpl<T> as a parameter type.

It’s rarely appropriate to use SmallVector<T, N> as a parameter type. If an API only reads from the vector, it should use ArrayRef.
```

I see in some places you have used SmallVectorImpl for the parameter but others like here just SmallVector. These should presumably all be either SmallVectorImpl or ArrayRef.

Comment (kazutakahirata) [reply llvm/lib/Analysis/ModuleSummaryAnalysis.cpp:104]: Thanks!  I'll go through the parameter types.  By the way, here, we have `SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &`.  We can cast `SmallVector<T, 0> &` to `SmallVectorImpl<T> &`, but I don't think we can cast `SetVector<ValueInfo, SmallVector<ValueInfo, 0>> &` to `SetVector<ValueInfo, SmallVectorImpl<ValueInfo>> &`.  Note that `SmallVector` is embedded in a parent type, namely `SetVector`.

**Comment by kazutakahirata** [comment]
Please take a look.  Thanks!

- I've updated the PR to use `SmallVector<ValueInfo> &` and `SmallVector<ValueInfo> &&` where I can.
- I've added a missing `std::move` to `ModuleSummaryIndexYAML.h`.

File: llvm/include/llvm/IR/ModuleSummaryIndex.h [hunk 2]
```diff
@@ -538,10 +538,11 @@ class GlobalValueSummary {
   /// (either by the initializer of a global variable, or referenced
   /// from within a function). This does not include functions called, which
   /// are listed in the derived FunctionSummary object.
-  std::vector<ValueInfo> RefEdgeList;
+  SmallVector<ValueInfo, 0> RefEdgeList;
```
Comment (minglotus-6) [inline llvm/include/llvm/IR/ModuleSummaryIndex.h:541]: nit: add a comment to explain the motivation to use `SmallVector<T, 0>` for future readers / maintainers.

**Review by mingmingl-llvm — APPROVED** [review]
lgtm as long as interface discussion (https://github.com/llvm/llvm-project/pull/107342/files#r1745875758) is resolved

**Review by teresajohnson — APPROVED** [review]
lgtm with mingming's comment suggestion

File: llvm/include/llvm/IR/ModuleSummaryIndex.h (hunk 2 above)
Comment (kazutakahirata) [reply llvm/include/llvm/IR/ModuleSummaryIndex.h:543]: Fixed in the latest iteration.  Thanks!

**Pull request merged**